### PR TITLE
feat(dnscheck): implement richer input

### PR DESCRIPTION
**Pull request by bassosimone**
This diff:

1. modifies `./internal/registry` and its `dnscheck.go` file such that `dnscheck` has its own private target loader;
2. modifies `./internal/targetloading` to allow reusing the code to read static input from CLI and from files;
3. modifies `./internal/model` to pass optional richer input to experiments;
4. modifies `./internal/engine` to pass richer input to experiments;
5. modifies `./internal/experiment/dnscheck` to read and use the richer input it is passed;
6. implements a custom target loader for `./internal/experiment/dnscheck` returning some static entries;
7. modifies `./internal/oonirun` to make sure richer-input experiments honor their options by moving option loading before the loading of targets.

Once this diff is merged, `miniooni` and `ooniprobe` will run `dnscheck` with richer input. Obviously, we are still lacking an API to fetch richer input, so for now this is static richer input, suitable as a starting point to land more diffs.

Part of https://github.com/ooni/probe/issues/2607.

### CODE REVIEW

File: internal/experiment/dnscheck/dnscheck.go [hunk 1]
```diff
@@ -114,7 +114,7 @@ func (m *Measurer) ExperimentVersion() string {
 // errors are in addition to any other errors returned by the low level packages
 // that are used by this experiment to implement its functionality.
 var (
-	ErrInputRequired        = errors.New("this experiment needs input")
+	ErrInputRequired        = targetloading.ErrInputRequired
```
Comment (bassosimone) [inline internal/experiment/dnscheck/dnscheck.go:117]: Note: this change breaks the ABI in terms of the exact error string that we emit in such a case. Because we do not have an API/ABI strict stability guarantee and because it makes sense to avoid having so many duplicate errors, I think it makes sense to apply this change in the interested of reducing unnecessary duplication a bit.

File: internal/oonirun/experiment_test.go [hunk 2]
```diff
@@ -70,10 +67,12 @@ func TestExperimentRunWithFailureToSubmitAndShuffle(t *testing.T) {
...
-								// Implementation note: the convention for input-less experiments is that
-								// they require a single entry containing an empty input.
-								entry := model.NewOOAPIURLInfoWithDefaultCategoryAndCountry("")
-								return []model.ExperimentTarget{entry}, nil
+								results := []model.ExperimentTarget{
+									model.NewOOAPIURLInfoWithDefaultCategoryAndCountry("a"),
+									model.NewOOAPIURLInfoWithDefaultCategoryAndCountry("b"),
+									model.NewOOAPIURLInfoWithDefaultCategoryAndCountry("c"),
+								}
+								return results, nil
```
Comment (bassosimone) [inline internal/oonirun/experiment_test.go:75]: This fixes a test that I broke previously. Now that we have a loader, we need to return "a", "b", "c" here rather than using the Inputs fields above. With this change we're back at 100% coverage.

File: internal/registry/factory_test.go [hunk 3]
```diff
@@ -860,3 +861,49 @@ func TestFactoryNewTargetLoaderWebConnectivity(t *testing.T) {
...
+func TestFactoryCustomTargetLoaderForRicherInput(t *testing.T) {
+	// create factory creating a custom target loader
+	factory := &Factory{
+		build:            nil,
+		canonicalName:    "",
+		config:           &customConfig{},
+		enabledByDefault: false,
+		inputPolicy:      "",
+		interruptible:    false,
+		newLoader: func(config *targetloading.Loader, options any) model.ExperimentTargetLoader {
```
Comment (DecFox) [inline internal/registry/factory_test.go:885]: I think we should have another test case similar to this one where we set `newLoader` as `nil` and assert that we get a valid loader from the passed in `ExperimentTargetLoaderConfig`

File: internal/reflectx/reflectx_test.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+	Age  int64
+	Ch   chan int64
+	F    bool
+	Fmp  map[string]*bool
+	Fp   *bool
+	Fv   []bool
+	Fvp  []*bool
+	Name string
+	Ptr  *int64
+	V    []int64
```
Comment (DecFox) [inline internal/reflectx/reflectx_test.go:20]:
```suggestion
	V    []int64
	p    string          
```
Although `_` is a private field, I think we should another private field explicitly to clarify that we only check for public fields using the `IsExported` check.

**Review by DecFox — CHANGES_REQUESTED** [review]
The changes look good to me. I left some comments which I think should make the diff good to go.

File: internal/registry/factory_test.go (hunk 3 above)
Comment (bassosimone) [reply internal/registry/factory_test.go]: ✔️

File: internal/reflectx/reflectx_test.go (hunk 4 above)
Comment (bassosimone) [reply internal/reflectx/reflectx_test.go]: ✔️

**Review by DecFox — APPROVED** [review]
LGTM!
